src/pal/tools/gen-buildsys-gcc.sh

```diff
@@ -80,55 +87,39 @@ for i in "$@"; do
     esac
 done
 
-OS=`uname`
+OS=$(uname)
 
 locate_gcc_exec() {
+  ENV_KNOB="CLR_$(echo "$1" | tr '[:lower:]' '[:upper:]')"
+  if [ -n "$(eval "\$$ENV_KNOB")" ]; then
```

Comment:
This change fixed it cross compilation. Your version was actually calling the compiler itself.

 locate_gcc_exec() {
   ENV_KNOB="CLR_$(echo "$1" | tr '[:lower:]' '[:upper:]')"
-  if [ -n "$(eval "\$$ENV_KNOB")" ]; then
-    eval "\$$ENV_KNOB"
+  if env | grep -q ^$ENV_KNOB=
+  then
+    eval "echo \"\$$ENV_KNOB\""

Reply:
Yup, unlike other branches which have `command -v`, here we want to `echo`. For readability, could it be rewritten like this:

```sh
ENV_KNOB="CLR_$(echo "$1" | tr '[:lower:]' '[:upper:]')"
KNOB_VALUE=$(eval "echo \"\$$ENV_KNOB\"")
if [ -n "$KNOB_VALUE" ]; then
  echo "$KNOB_VALUE"
  return
fi
```

Reply:
Variable detection was also not working. That's why, I ended up changing with grep -q.

Reply:
Can you please try both the cross target and the x64 target on this branch (without other changes)?

On the tip of my branch (`d3eb8f5`), using Ubuntu I applied:

```diff
 if ! gcc_ar="$(locate_gcc_exec ar)"; then { echo "Unable to locate gcc-ar"; exit 1; } fi
+echo $gcc_ar
+exit 1

 if ! gcc_nm="$(locate_gcc_exec nm)"; then { echo "Unable to locate gcc-nm"; exit 1; } fi
```

and tested it like this:

```sh
rm -rf bin ; CLR_AR=/custom/path/to/ar ./build.sh x64 checked gcc
# prints /custom/path/to/ar
rm -rf bin ; ./build.sh x64 checked gcc
# prints /usr/bin/ar
```

Reply:
Still not working. it is still trying to compile x86_64.

I think this has costed more time than necessary for a cosmetic change. I have been debugging this whole night.

Please restore the compiler detection since these cosmetic changes are not necessary.

I don't think I can afford any more time on this.

Reply:
> Still not working

Can you share some information about your setup so someone else can investigate / reproduce? On my box, it is working. Like which operating system and version have you tried it on?

> a cosmetic change

Master is failing with:

```
Detected Linux x86_64
CMake Error at src/corefx/System.Globalization.Native/CMakeLists.txt:19 (message):
  Cannot find libicuuc, try installing libicu-dev (or the appropriate package
  for your platform)


-- Configuring incomplete, errors occurred!
```
PR fixes it.
The main difference is that in master lttng-ust and libicuuc cmake introspection fails, because the value of `gcc_link` is incorrect for x64 target. It has to use `link`.

Reply:
Let's focus on lttng fix. I have no issues with that side of the change.
I don't think any of the changes in gcc script are necessary though..

You can use an arm64 GCC cross-compiler and use it aganst ubuntu arm64 OS using sysroot argument.

OS detection is not working with your change set.
It is detecting x86_64 Linux with your set when cross compiling and arm64 Linux without.

Original build instructions:

export CC="aarch64-gnu-linux-gcc  --sysroot=mysysrootdir"
export CXX="aarch64-gnu-linux-g++  --sysroot=mysysrootdir"
export LD="aarch64-gnu-linux-ld  --sysroot=mysysrootdir"

./build.sh -gcc -cross -arm64 -ignorewarnings  -skipcrossarchnative -skipmanaged

Possible places to look:
configurecompiler.cmake:      message("Detected Linux x86_64")
configurecompiler.cmake:      message("Detected Linux ARM")
configurecompiler.cmake:      message("Detected Linux ARM64")
configurecompiler.cmake:      message("Detected Linux i686")
configurecompiler.cmake:    message("Detected OSX x86_64")
configurecompiler.cmake:    message("Detected FreeBSD amd64")
configurecompiler.cmake:    message("Detected NetBSD amd64")

Reply:
how about just fixing "the value of gcc_link is incorrect for x64 target. It has to use link." with a single line change?

Reply:
Single line change will not help here, as first the script (in master branch) sets $CC idempotently: https://github.com/dotnet/coreclr/blob/24d05f5e6c/src/pal/tools/gen-buildsys-gcc.sh#L48, then later it tests for linker: https://github.com/dotnet/coreclrhttps://github.com/dotnet/coreclr/blob/24d05f5e6c/src/pal/tools/gen-buildsys-gcc.sh#L100.

I have tried your previous approach to use grep, and it seems to satisfy multiple platforms. Change is now pushed.